rust/template/differential_datalog/profile.rs

```diff
@@ -38,34 +38,92 @@ pub enum ProfMsg {
 
 #[derive(Clone)]
 pub struct Profile {
-    names: FnvHashMap<usize, String>,
-    sizes: FnvHashMap<usize, isize>,
-    peak_sizes: FnvHashMap<usize, isize>
+    addresses:  SequenceTrie<usize, usize>,
+    names:      FnvHashMap<usize, String>,
+    sizes:      FnvHashMap<usize, isize>,
+    peak_sizes: FnvHashMap<usize, isize>,
+    starts:     FnvHashMap<(usize,usize), Duration>,
+    durations:  FnvHashMap<usize, Duration>
 }
 
 impl fmt::Display for Profile {
     fn fmt(&self, f: &mut fmt::Formatter) -> Result<(), fmt::Error> {
-        let res1: Result<(), fmt::Error> = self.sizes.iter().map(|(operator, size)| {
+        write!(f, "\nArrangement size profile\n")?;
+        let mut sizes: Vec<(usize, isize)> = self.sizes.clone().into_iter().collect();
+        sizes.sort_by(|a,b| a.1.cmp(&b.1).reverse());
+        let current: Result<(), fmt::Error> = sizes.iter().map(|(operator, size)| {
             let name = self.names.get(operator).map(|s|s.as_ref()).unwrap_or("???");
-            write!(f, "current size of {} {}: {}\n", name, operator, size)
+            let msg = format!("{} {}:", name, operator);
+            write!(f, "{: <80} {}\n", msg, size)
         }).collect();
-        res1.and(
-            self.peak_sizes.iter().map(|(operator, size)| {
-                let name = self.names.get(operator).map(|s|s.as_ref()).unwrap_or("???");
-                write!(f, "peak size of {} {}: {}\n", name, operator, size)
-            }).collect())
+        current?;
+
+        write!(f, "\nArrangement peak sizes\n")?;
+        let mut peak_sizes: Vec<(usize, isize)> = self.peak_sizes.clone().into_iter().collect();
```

Comment:
this code is common, perhaps it can be factored in a separate function?

Reply:
fixed